app/controllers/solidus_stripe/intents_controller.rb

```diff
@@ -0,0 +1,83 @@
+# frozen_string_literal: true
+
+require 'stripe'
+
+class SolidusStripe::IntentsController < Spree::BaseController
+  include Spree::Core::ControllerHelpers::Order
+
+  before_action :load_payment_method
+
+  def payment_confirmation
+    intent = @payment_method.gateway.request { Stripe::PaymentIntent.retrieve(id: params[:payment_intent]) }
+    payment = @payment_method.find_in_progress_payment_for(current_order)
+
+    unless %w[confirm payment].include?(current_order.state.to_s)
+      redirect_to main_app.checkout_state_path(current_order.state)
+      return
+    end
+
+    current_order.state = :payment
+    # https://stripe.com/docs/payments/intents?intent=payment
+    case intent.status
+    when 'requires_payment_method'
+      # this state after an attempt at processing a payment means that it failed.
+      payment.failure! unless payment.failure?
+      ensure_state_is(payment, :processing)
+    when 'requires_confirmation'
+      ensure_state_is(payment, :processing)
```

Comment:
Aren't those `ensure_state_is` defensive? Shouldn't our logic ensure that the state is the expected one?

Reply:
Yes they are, and that's because of how many moving parts and configuration options we have in Solidus. Eventually when we'll have better coverage we might want to relax that a bit. 

But they're also very lightweight so maybe the work as runnable documentation? Anyway I think we can decide later